File: nfc_proof_of_concept/packages/nfc_host_card_emulation/android/src/main/kotlin/io/flutter/plugins/nfc_host_card_emulation/app_layer/file_access/fields/capability_container_fields.kt
package io.flutter.plugins.nfc_host_card_emulation.file_access.fields

import io.flutter.plugins.nfc_host_card_emulation.app_layer.ApduField

class CcLenField(len: Int) : ApduField("CCLEN", 2) {
    companion object {
        val defaultLen = CcLenField(15)
    }

    init {
        require(len in 15..0xFFFF) { "Capability Container length must be >= 15." }
        buffer[0] = (len shr 8 and 0xFF).toByte()
        buffer[1] = (len and 0xFF).toByte()
    }
}

class CcMappingVersionField private constructor(version: Int) : ApduField("MappingVersion", 1) {
    companion object {
        val v2_0 = CcMappingVersionField(0x20)
    }

    init {
        buffer[0] = version.toByte()
    }
}

class CcMaxApduDataSizeField private constructor(size: Int, name: String) : ApduField(name, 2) {
    companion object {
        fun mLe(size: Int = 0x00FF) = CcMaxApduDataSizeField(size, "MLe")
        fun mLc(size: Int = 0x00FF) = CcMaxApduDataSizeField(size, "MLc")
    }

    init {
        require(size in 1..0xFFFF) { "$name size must be a positive 16-bit integer." }
        buffer[0] = (size shr 8 and 0xFF).toByte()
        buffer[1] = (size and 0xFF).toByte()
    }
}

File: nfc_proof_of_concept/packages/flutter_hce/lib/app_layer/file_access/fields/capability_container_fields.dart
import '../../field.dart';

class CcLenField extends ApduField {
  static final defaultLen = CcLenField(15);

  CcLenField(int len) : super(size: 2, name: "CCLEN") {
    if (len < 15 || len > 0xFFFF) {
      throw ArgumentError('Capability Container length must be >= 15.');
    }
    buffer[0] = (len >> 8) & 0xFF;
    buffer[1] = len & 0xFF;
  }
}

class CcMappingVersionField extends ApduField {
  static final v2_0 =
      CcMappingVersionField._internal(0x20, name: "MappingVersion (2.0)");

  CcMappingVersionField._internal(int version, {required String name})
      : super(size: 1, name: name) {
    buffer[0] = version;
  }

  /// Smart constructor that tries to return existing static final instances first
  factory CcMappingVersionField.fromByte(int version, {String? name}) {
    // Try to match with existing static final instances
    if (version == 0x20) {
      return v2_0;
    }

    // If no match found, create new instance
    final majorVersion = (version >> 4) & 0x0F;
    final minorVersion = version & 0x0F;
    final effectiveName =
        name ?? "MappingVersion ($majorVersion.$minorVersion)";
    return CcMappingVersionField._internal(version, name: effectiveName);
  }
}

class CcMaxApduDataSizeField extends ApduField {
  static final defaultMLe = CcMaxApduDataSizeField.mLe(0x00FF);
  static final defaultMLc = CcMaxApduDataSizeField.mLc(0x00FF);

  CcMaxApduDataSizeField.mLe(int size) : super(size: 2, name: "MLe") {
    if (size <= 0 || size > 0xFFFF) {
      throw ArgumentError('MLe size must be a positive 16-bit integer.');
    }
    buffer[0] = (size >> 8) & 0xFF;
    buffer[1] = size & 0xFF;
  }
  CcMaxApduDataSizeField.mLc(int size) : super(size: 2, name: "MLc") {
    if (size <= 0 || size > 0xFFFF) {
      throw ArgumentError('MLc size must be a positive 16-bit integer.');
    }
    buffer[0] = (size >> 8) & 0xFF;
    buffer[1] = size & 0xFF;
  }
}
